Packet++/header/GvcpLayer.h

```diff
@@ -0,0 +1,626 @@
+#pragma once
+
+#include "IpAddress.h"
+#include "Layer.h"
+#include "MacAddress.h"
+#include "SystemUtils.h"
+#include <iostream>
+
+/**
+ * @file GvcpLayer.h
+ * @author An-Chi Liu (phy.tiger@gmail.com)
+ */
+
+/**
+ * @namespace pcpp
+ * @brief The main namespace for the PcapPlusPlus lib
+ */
+namespace pcpp
+{
+	namespace internal
+	{
+		static constexpr size_t kGvcpMagicNumber = 0x42;
+		static constexpr size_t kGvcpRequestHeaderLength = 8;
+		static constexpr size_t kGvcpAckHeaderLength = 8;
+		static constexpr size_t kGvcpDiscoveryBodyLength = 248;
+		static constexpr size_t kGvcpForceIpBodyLength = 56;
+	}  // namespace internal
+
+	using GvcpFlag = uint8_t;  // flag bits are specified by each command
+
+	/// @brief Gvcp command
+	/// See spec "18 Command and Acknowledge Values"
+	enum class GvcpCommand : uint16_t
+	{
+		// Discovery Protocol Control
+		DiscoveredCmd = 0x0002,
+		DiscoveredAck = 0x0003,
+		ForceIpCmd = 0x0004,
+		ForceIpAck = 0x0005,
+
+		// Streaming Protocol Control
+		PacketResendCmd = 0x0040,
+		PacketResendAck = 0x0041,  // Resent packet must be on the stream channel
+
+		// Device Memory Access
+		ReadRegCmd = 0x0080,
+		ReadRegAck = 0x0081,
+		WriteRegCmd = 0x0082,
+		WriteRegAck = 0x0083,
+		ReadMemCmd = 0x0084,
+		ReadMemAck = 0x0085,
+		WriteMemCmd = 0x0086,
+		WriteMemAck = 0x0087,
+		PendingAck = 0x0089,
+
+		// Asynchronous Events
+		EventCmd = 0x00C0,
+		EventAck = 0x00C1,
+		EventDataCmd = 0x00C2,
+		EventDataAck = 0x00C3,
+
+		// Miscellaneous
+		ActionCmd = 0x0100,
+		ActionAck = 0x0101,
+		Unknown = 0xFFFF
+	};
+
+	std::ostream& operator<<(std::ostream& os, GvcpCommand command);
+
+	/// @brief Gvcp response status
+	/// See spec "Table 19-1: List of Standard Status Codes"
+	enum class GvcpResponseStatus : uint16_t
+	{
+		Success = 0x0000,
+		PacketResend = 0x0100,
+		NotImplemented = 0x8001,
+		InvalidParameter = 0x8002,
+		InvalidAddress = 0x8003,
+		WriteProtect = 0x8004,
+		BadAlignment = 0x8005,
+		AccessDenied = 0x8006,
+		Busy = 0x8007,
+		LocalProblem = 0x8008,     // deprecated
+		MsgMismatch = 0x8009,      // deprecated
+		InvalidProtocol = 0x800A,  // deprecated
+		NoMsg = 0x800B,            // deprecated
+		PacketUnavailable = 0x800C,
+		DataOverrun = 0x800D,
+		InvalidHeader = 0x800E,
+		WrongConfig = 0x800F,  // deprecated
+		PacketNotYetAvailable = 0x8010,
+		PacketAndPrevRemovedFromMemory = 0x8011,
+		PacketRemovedFromMemory = 0x8012,
+		NoRefTime = 0x8013,                     // GEV 2.0
+		PacketTemporarilyUnavailable = 0x8014,  // GEV 2.0
+		Overflow = 0x8015,                      // GEV 2.0
+		ActionLate = 0x8016,                    // GEV 2.0
+		LeaderTrailerOverflow = 0x8017,         // GEV 2.1
+		Error = 0x8FFF,
+		Unknown = 0xFFFF
+	};
+
+	std::ostream& operator<<(std::ostream& os, GvcpResponseStatus status);
+
+#pragma pack(push, 1)
+	/// @brief Gvcp request header
+	/// @note refer to the spec "15.1 Request Header". The data is stored as big-endian.
+	struct GvcpRequestHeader
```

Comment:
I have a general design comment: I think it'd be nicer to add the property getters / setters to the layer class and not to the struct. That's the way we usually design layers. We can do something like this (below is pseudo code):
```cpp
class GvcpAcknowledgeLayer : public GvcpLayer
{
    uint16_t getAckId() const { return getGvcpHeader()->ackId; }
}
```

Reply:
How about having `getAckId` for both the layer and the header so that a user can get the data either way?

Reply:
I think that we can "hide" the struct, and only expose the getters / setters. We can make `getGvcpHeader()` private / protected

Reply:
@tigercosmos this was not changed. Please see other protocol implementations like the recent `WireGuardLayer` or `BgpLayer`

Reply:
@seladb  please read the implementation in the layer class. I intend to keep the function for the header. The layer will use the functions from the header. In this way, user can get the field in either the header or the layer.

Reply:
I think it makes it harder to use and differs from other implementations we have in our code base.

When a user has a layer instance, for example `GvcpRequestHeader`, they expect to have direct access to the getters and setters:
```cpp
auto gvcpLayer = ...;
gvcpLayer->getFlag();
```

Instead what they need to do with the current implementation is:
```cpp
auto gvcpLayer = ...;
gvcpLayer->getGvcpHeader->getFlag();
```

I think the first implementation is more clear, and aligns better to other protocols

Reply:
@seladb  I think you misunderstand my statement. I meant there are two ways to get the flag value in your example.

```cpp
auto gvcpLayer = .... ;
gvcpLayer->getFlag() == gvcpLayer->getGvcpHeader->getFlag();
```

Reply:
I can't see the first option `gvcpLayer->getFlag()` in the code base, am I missing something? 🤔 

Also - why do we need to have both? Isn't it enough to keep the getters and setters in the layer?

Reply:
An example is here: https://github.com/seladb/PcapPlusPlus/pull/1357#discussion_r1784026612

I thought it's more convenient if the header and the layer both are able to get the field, but I guess it's fine to keep the layer's one.

Reply:
@seladb  I found that there is a reason that at least `GvcpRequestHeader`  and `GvcpAckHeader` need to have `getCommand()` method. When we parse the layer, I first check the header and decide which layer to create:

```cpp
GvcpLayer* GvcpLayer::parseGvcpLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
{
	if (GvcpLayer::verifyRequest(data))
	{	
		auto* header = reinterpret_cast<GvcpRequestHeader*>(data);
		if (header->getCommand() == GvcpCommand::DiscoveredCmd)
		{
			return new GvcpDiscoveryRequestLayer(data, dataLen, prevLayer, packet);
		}
		else if (header->getCommand() == GvcpCommand::ForceIpCmd)
		{
			return new GvcpForceIpRequestLayer(data, dataLen, prevLayer, packet);
		}
		else
		{
			return new GvcpRequestLayer(data, dataLen, prevLayer, packet);
		}
	}
	else
	{
		auto* header = reinterpret_cast<GvcpAckHeader*>(data);
		if (header->getCommand() == GvcpCommand::DiscoveredAck)
		{
			return new GvcpDiscoveryAcknowledgeLayer(data, dataLen);
		}
		else if (header->getCommand() == GvcpCommand::ForceIpAck)
		{
			return new GvcpForceIpAcknowledgeLayer(data, dataLen);
		}
		else
		{
			return new GvcpAcknowledgeLayer(data, dataLen, prevLayer, packet);
		}
	}
}
```

Of course we can do something like the following code, but it's inefficient:
```cpp
auto layer = new GvcpRequestLayer(data, dataLen, prevLayer, packet);
if (layer->getCommand() == GvcpCommand::DiscoveredCmd)
{
	delete layer;
	return  new GvcpDiscoveryRequestLayer(data, dataLen, prevLayer, packet);
}
```

What do you think?

Reply:
Sure.